crates/core/app/tests/common/mod.rs

```diff
@@ -83,19 +87,134 @@ impl TempStorageExt for TempStorage {
+                    .tap(|overwritten| {
+                        // Log a warning if this overwrote any validators already in the app state.
+                        warn!(?overwritten, "`with_penumbra_auto_app_state` overwrote validators in the given AppState")
+                    });
+                AppState::Content(content)
+            }
+        };
+
+        // Serialize the app state into bytes, and add it to the builder.
+        serde_json::to_vec(&app_state)
+            .map_err(Self::Error::from)
+            .map(|s| self.app_state(s))
+    }
+
+    fn with_penumbra_single_validator(self) -> Self {
+        use {
+            decaf377_rdsa::VerificationKey,
+            penumbra_keys::keys::{SpendKey, SpendKeyBytes},
+            penumbra_proto::{
+                core::keys::v1::{GovernanceKey, IdentityKey},
+                penumbra::core::component::stake::v1::Validator,
+            },
+            rand::Rng,
+            rand_core::OsRng,
+        };
+
+        // Generate a spend authoration key.
+        let bytes = {
+            let spend_key = SpendKey::from(SpendKeyBytes(OsRng.gen()));
+            let spend_auth_key = spend_key.spend_auth_key();
+            let verification_key = VerificationKey::from(spend_auth_key);
+            verification_key.to_bytes().to_vec()
+        };
+
+        // Generate an identity key for the validator.
+        let identity_key = Some(IdentityKey { ik: bytes.clone() });
+
+        // Generate a governance key for the validator.
+        //
+        // NB: for now, we will use the same key for governance. See the documentation of
+        // `GovernanceKey` for more information about cold storage of validator keys.
+        let governance_key = Some(GovernanceKey { gk: bytes.clone() });
+
+        // Generate a consensus pubkey for the validator.
+        let consensus_key = {
+            let signing_key = ed25519_consensus::SigningKey::new(OsRng);
+            signing_key.as_bytes().as_slice().to_vec()
+        };
+
+        let validator = Validator {
+            identity_key,
+            consensus_key,
+            enabled: true,
+            sequence_number: 0,
+            governance_key,
+            name: String::default(),
+            website: String::default(),
+            description: String::default(),
+            funding_streams: Vec::default(),
+        };
+
+        Self {
+            validators: vec![validator],
+            ..self
+        }
+    }
+}
+
+pub trait TestNodeExt {
+    fn penumbra_identity_key(&self) -> penumbra_stake::IdentityKey;
+}
+
+impl<C> TestNodeExt for TestNode<C> {
+    //  TODO(kate): `penumbra_stake::IdentityKey` could be `From<[u8; 32]>`
+    fn penumbra_identity_key(&self) -> penumbra_stake::IdentityKey {
+        use penumbra_proto::core::keys::v1 as pb;
+        use prost::Message;
+        pb::IdentityKey::decode(self.identity_key().as_slice())
+            .unwrap()
+            .try_into()
+            .unwrap()
```

Comment:
this is where i am getting a panic rn,

`called `Result::unwrap()` on an `Err` value: DecodeError { description: "invalid tag value: 0", stack: [] }`

i need to do this dance slightly differently, probably by referring directly to the `decaf377` type instead of using `prost::Message`. i'll figure that out tomorrow!